framework/src/modules/bft/bft_params.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright © 2021 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { codec } from '@liskhq/lisk-codec';
+import { BIG_ENDIAN, intToBuffer } from '@liskhq/lisk-cryptography';
+import { ImmutableSubStore } from '../../node/state_machine';
+import { SubStore } from '../../node/state_machine/types';
+import { BFTParameterNotFoundError } from './errors';
+import { BFTParameters, bftParametersSchema } from './schemas';
+
+export const getBFTParameters = async (
+	paramsStore: ImmutableSubStore,
+	height: number,
+): Promise<BFTParameters> => {
+	const start = intToBuffer(0, 4, BIG_ENDIAN);
+	const end = intToBuffer(height, 4, BIG_ENDIAN);
+	const results = await paramsStore.iterate({
+		limit: 1,
+		start,
+		end,
+		reverse: true,
+	});
+	if (results.length !== 1) {
+		throw new BFTParameterNotFoundError();
+	}
+	const [result] = results;
+	return codec.decode<BFTParameters>(bftParametersSchema, result.value);
+};
+
+export const deleteBFTParameters = async (paramsStore: SubStore, height: number): Promise<void> => {
+	const exist = await paramsStore.has(intToBuffer(height, 4, BIG_ENDIAN));
+	if (!exist) {
+		return;
+	}
+	const start = intToBuffer(0, 4, BIG_ENDIAN);
+	const end = intToBuffer(Math.max(height - 1, 0), 4, BIG_ENDIAN);
+	const results = await paramsStore.iterate({
+		start,
+		end,
+	});
+	for (const result of results) {
+		await paramsStore.del(result.key);
+	}
+};
+
+export class BFTParametersCache {
```

Comment:
I think this cache is a good idea to avoid many reads. As far as I understand though, I have a separate cached value for every height. This means if in the Lisk Mainchain I iterate over 309 blocks (for instance in `updateMaxHeightPrevoted`), then I would call the the catch for 309 not yet cached values which result in 309 calls to `getBFTParameters`. This results in 309 iterations on the state store.

However, the state store only stores one key-value pair per round so with one of these iterations I could in theory add 103 items to the cache. This would reduce the number of calls to the state store to 3-4 (4 as rounds 309 blocks may cover 4 rounds).

Reply:
@shuse2 Do you think some further optimization as suggested above is possible and would make sense to introduce?

Reply:
wouldnt it difficult to predict how frequent `setBFTParameters` will be called? (we know how frequent it will be called for DPoS, but not for other modules)

Reply:
As far as I understand, the cache is only used for in the `afterBlockExecute` hook. During this hook, the BFT Parameters store stays unchanged and I only use the BFT Parameters for heights that appear in `bftVotes.blockBFTInfos`.  

So what I could do is one iteration of the state store to create a cache for the height `bftVotes.blockBFTInfos[-1].height` to `bftVotes.blockBFTInfos[0].height`.

Reply:
Nice solution for the cache now. I resolve this comment.